tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java

```diff
@@ -0,0 +1,269 @@
+import tla2sany.semantic.Context;
+import tla2sany.semantic.Errors;
+import tla2sany.semantic.FormalParamNode;
+import tla2sany.semantic.Generator;
+import tla2sany.semantic.LetInNode;
+import tla2sany.semantic.LevelConstants;
+import tla2sany.semantic.ModuleNode;
+import tla2sany.semantic.OpApplNode;
+import tla2sany.semantic.OpArgNode;
+import tla2sany.semantic.OpDefNode;
+import tla2sany.semantic.SemanticNode;
+import tla2sany.semantic.SymbolNode;
+import tla2sany.st.Location;
+import tlc2.tool.impl.SpecProcessor;
+import util.ToolIO;
+
+/**
+ * An expression for conditionally triggering a breakpoint.
+ */
+public class TLCBreakpointExpression {
+
+	/**
+	 * Given a spec and an unparsed expression, build an operator that can be
+	 * evaluated within the context of that spec.
+	 *
+	 * @param processor     Processes expression within the model context.
+	 * @param semanticRoot  The root of the spec's semantic parse tree.
+	 * @param location      The breakpoint location.
+	 * @param conditionExpr The unparsed breakpoint expression.
+	 * @return A breakpoint expression, or null if parsing failed.
+	 */
+	public static OpDefNode process(
+			final SpecProcessor processor,
+			final ModuleNode semanticRoot,
+			final Location location,
+			final String conditionExpr
+	) {
+		if (null == conditionExpr || conditionExpr.isBlank()) {
+			return null;
+		}
+		ToolIO.out.println("BPExpr: processing expression \"" + conditionExpr + "\"");
+
+		final String rootModName = semanticRoot.getName().toString();
+		final String bpModName = generateUnusedName(semanticRoot, "__BreakpointModule__%s");
+		ToolIO.out.println("BPExpr: wrapping with module \"" + bpModName + "\"");
+		final String bpOpName = generateUnusedName(semanticRoot, "__BreakpointExpr__%s");
+		final Set<String> paramNames = getScopedIdentifiers(semanticRoot, location);
+		final String params = paramNames.size() > 0 ? "(" + String.join(", ", paramNames) + ")" : "";
+		final String bpOpDef = bpOpName + params;
+		ToolIO.out.println("BPExpr: wrapping with op \"" + bpOpDef + "\"");
+
+		final String wrapper = "---- MODULE %s ----\nEXTENDS %s\n%s == %s\n====";
+		String wrappedConditionExpr = String.format(wrapper, bpModName, rootModName, bpOpDef, conditionExpr);
+		byte[] wrappedConditionExprBytes = wrappedConditionExpr.getBytes(StandardCharsets.UTF_8);
+		InputStream sourceCode = new ByteArrayInputStream(wrappedConditionExprBytes);
+		TLAplusParser parser = new TLAplusParser(sourceCode, StandardCharsets.UTF_8.name());
+		boolean syntaxParseSuccess = parser.parse();
+		SyntaxTreeNode syntaxRoot = parser.ParseTree;
+		if (!syntaxParseSuccess || null == syntaxRoot) {
+			// Parse error is output to ToolIO.out
+			return null;
+		}
+		ToolIO.out.println("BPExpr: success [syntactic analysis]");
+
+		Errors semanticLog = new Errors();
+		SemanticNode.setError(semanticLog); // Annoyingly static
+		Generator semanticChecker = new Generator(semanticRoot.semanticChecker.moduleTable, semanticLog);
```

Comment:
There is no need to change `ModuleNode` (see above) because this can use `SpecProcessor#getModuleTbl`.